satellite/metainfo/loop.go

```diff
@@ -174,7 +175,61 @@ waitformore:
 		}
 	}
 
-	err = loop.db.Iterate(ctx, storage.IterateOptions{Recurse: true},
+	return iterateDatabase(ctx, loop.db, observers)
+}
+
+// handlePointer deals with a pointer for a single observer
+// if there is some error on the observer, handle the error and return false. Otherwise, return true
+func handlePointer(ctx context.Context, observer *observerContext, path ScopedPath, isLastSegment bool, pointer *pb.Pointer) bool {
+	switch pointer.GetType() {
+	case pb.Pointer_REMOTE:
+		if observer.HandleError(observer.RemoteSegment(ctx, path, pointer)) {
+			return false
+		}
+	case pb.Pointer_INLINE:
+		if observer.HandleError(observer.InlineSegment(ctx, path, pointer)) {
+			return false
+		}
+	default:
+		return false
+	}
+	if isLastSegment {
+		if observer.HandleError(observer.Object(ctx, path, pointer)) {
+			return false
+		}
+	}
+
+	select {
+	case <-observer.ctx.Done():
+		observer.HandleError(observer.ctx.Err())
+		return false
+	default:
+	}
+
+	return true
+}
+
+// Wait waits for run to be finished.
+// Safe to be called concurrently.
+func (loop *Loop) Wait() {
+	<-loop.done
+}
+
+// IterateDatabase iterate over PointerDB and notify specified observers about results
+func IterateDatabase(ctx context.Context, db PointerDB, observers ...Observer) error {
```

Comment:
nit: Keep this above `handlePointer` because it's public and uses handle pointer indirectly.